csr: add expirationSeconds field to control cert lifetime

enj opened the pull request:
This change updates the CSR API to add a new, optional field called
expirationSeconds.  This field is a request to the signer for the
maximum duration the client wishes the cert to have.  The signer is
free to ignore this request based on its own internal policy.  The
signers built-in to KCM will honor this field if it is not set to a
value greater than --cluster-signing-duration.  The minimum allowed
value for this field is 600 seconds (ten minutes).

This change will help enforce safer durations for certificates in
the Kube ecosystem and will help related projects such as
cert-manager with their migration to the Kube CSR API.

Future enhancements may update the Kubelet to take advantage of this
field when it is configured in a way that can tolerate shorter
certificate lifespans with regular rotation.

Signed-off-by: Monis Khan <mok@vmware.com>

/kind feature
/kind api-change
/sig auth
/priority important-longterm
/triage accepted
@kubernetes/sig-auth-pr-reviews @kubernetes/sig-security-pr-reviews
/assign @mikedanese @liggitt @munnerz 

Fixes #92678

xref: https://github.com/jetstack/cert-manager/pull/3646

```release-note
The CertificateSigningRequest.certificates.k8s.io API supports an optional expirationSeconds field to allow the client to request a particular duration for the issued certificate.  The default signer implementations provided by the Kubernetes controller manager will honor this field as long as it does not exceed the --cluster-signing-duration flag.
```

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2784
- [Usage]: https://github.com/kubernetes/website/pull/28070
```

#### Special notes for your reviewer:

- There are several TODOs in regards to minor details

enj commented on staging/src/k8s.io/api/testdata/v1.19.0/certificates.k8s.io.v1.CertificateSigningRequest.after_roundtrip.json:
```diff
@@ -0,0 +1,74 @@
+{
```
These changes look wrong to me but they are what `UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test k8s.io/api` auto generated.  I assume I have a mistake somewhere.

liggitt commented on staging/src/k8s.io/api/certificates/v1/types.go:
```diff
@@ -84,6 +84,11 @@ type CertificateSigningRequestSpec struct {
+	// notAfterHint is a hint to the signer in regards to when the issued certificate should expire.
+	// The signer may or may not honor this field.  The well-known kubernetes signers will honor this field
+	// as long as the requested notAfter time is not later than the maximum notAfter time they will honor.
+	NotAfterHint metav1.Time `json:"notAfterHint,omitempty" protobuf:"bytes,8,opt,name=notAfterHint"`
```
You have to make this a pointer for it to actually be omitted when empty. This will remove the need for the after_roundtrip changes

enj commented:
/milestone v1.22

awly commented:
/uncc awly

mikedanese commented on pkg/apis/certificates/types.go:
```diff
@@ -68,6 +68,11 @@ type CertificateSigningRequestSpec struct {
+	// durationHint is a hint to the signer in regards to when the issued certificate should expire.
+	// The signer may or may not honor this field.  The well-known kubernetes signers will honor this field
+	// as long as the requested duration is not later than the maximum duration they will honor.
+	DurationHint *metav1.Duration
```
I wounder if something like `RequestedValidityDuration` or `RequestedTTL` would be clearer.

enj replied:
I just went with `durationHint` because of what I recalled from the SIG meetings.

liggitt replied:
per [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units), include the unit in the field name and make it an int32

liggitt replied:
Hint is a first in a field name... anything in spec is implicitly "desired state". 

DurationSeconds or something similar is fine, and document it may or may not be honored by the issuer.

For prior art, see the token expiration request field:

```
// ExpirationSeconds is the requested duration of validity of the request. The
// token issuer may return a token with a different validity duration so a
// client needs to check the 'expiration' field in a response.
// +optional
ExpirationSeconds *int64 `json:"expirationSeconds" protobuf:"varint,4,opt,name=expirationSeconds"`
```

liggitt replied:
Also, document known cases where the issuer would not honor it:

1. old signer that's unaware of the field
2. signer whose configured max is shorter than the requested duration
3. signer whose configured min is longer than the requested duration

mikedanese commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +32,79 @@ import (
+	Backdate time.Duration
+
+	// Short is the duration used to determine if the lifetime of a certificate should be considered short.
+	Short time.Duration
```
Can this be a package level const until we need to expose it?

enj replied:
I exposed it mostly to be consistent with `Backdate`.  AFAICT they are only set in unit tests so I can make them both private if you prefer.

enj replied:
Err actually the signer is in a different package so we would need to hardcode all values which makes this a bit harder to unit test.

mikedanese commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +32,79 @@ import (
+	Short time.Duration
+
+	// Now defaults to time.Now but can be stubbed for testing
+	Now func() time.Time
```
Why not `time.Time` instead of `func() time.Time`

enj replied:
The func based approach is what `CertificateAuthority` used.  I can make either way work if you have a preference.

mikedanese commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -185,15 +189,12 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
-		TTL:    s.certTTL,
-		Usages: usages,
-	})
+	der, err := currCA.Sign(x509cr.Raw, s.policy, usages, durationHint)
```
`authority.PermissiveSigningPolicy` is intended to be used as named arguments to Sign. Can you preserve that API?

mikedanese replied:
It is a container of arguments so that the call site is clear and arguments can be added without breaking API compatibility. Can we keep the package API as is?

mikedanese commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -31,40 +32,79 @@ import (
+	// NotAfter value of the certificate.  Backdate is ignored when TTL is less than Short.
 	TTL time.Duration
-	// Usages are the allowed usages of a certificate.
-	Usages []capi.KeyUsage
```
Why are we moving this?

enj replied:
The structure and comment did not make sense together because we simply pass in the usages from the CSR.  So it is not allowed usages but "validated usage strings" or something like that.

mikedanese replied:
https://github.com/kubernetes/kubernetes/blob/7705b300e2085c3864bb1e49a7302bf17f080219/staging/src/k8s.io/api/certificates/v1beta1/generated.proto#L104-L107

Allowed usages are a property of a certificate. The thing that validates allowed usages is a verifier of the certificate. I admit that the godoc is not super helpful, but it seems accurate.

mikedanese commented on staging/src/k8s.io/client-go/util/certificate/certificate_manager.go:
```diff
@@ -88,6 +88,8 @@ type Config struct {
 	// generated by the manager.
 	SignerName string
+	// Duration is the requested lifetime length for certificates generated by the manager.
+	Duration time.Duration
```
Can you give this a clearer name?

enj replied:
`RequestedCertificateLifetime`?

mikedanese replied:
I like it.

mikedanese commented on staging/src/k8s.io/client-go/util/certificate/csr/csr.go:
```diff
@@ -62,6 +62,9 @@ func RequestCertificate(client clientset.Interface, csrData []byte, name string,
 	if len(csr.Name) == 0 {
 		csr.GenerateName = "csr-"
 	}
+	if duration != 0 {
```
Should we attempt to match apiserver validation? i.e. >= min?

enj replied:
You get a nice validation error from the server if you pass in something that does not make sense?

mikedanese reviewed (commented):
Sweet integration test.

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -205,6 +206,10 @@ func validateCertificateSigningRequest(csr *certificates.CertificateSigningReque
 	}
+	const min = 5 * time.Minute
+	if csr.Spec.DurationHint != nil && csr.Spec.DurationHint.Duration < min {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("durationHint"), csr.Spec.DurationHint.Duration, "may not specify a duration less than 5 minutes"))
```
What was the rationale for a minimum of 5 minutes?

liggitt replied:
For reference, the projected token lifetime minimum is 10 minutes, and kubelet renewals start at 80% of that, so 8 minutes. Given CSR approval/signing is async, I'd expect to leave at least as much time for a CSR renewal request to be processed... starting renewal 3 minutes after issuance seems aggressive.

liggitt commented on staging/src/k8s.io/client-go/util/certificate/csr/csr.go:
```diff
@@ -46,7 +46,7 @@ import (
 // before but not to completion) or create a certificate signing request using the
 // PEM encoded CSR and send it to API server.
-func RequestCertificate(client clientset.Interface, csrData []byte, name string, signerName string, usages []certificatesv1.KeyUsage, privateKey interface{}) (reqName string, reqUID types.UID, err error) {
+func RequestCertificate(client clientset.Interface, csrData []byte, name, signerName string, duration time.Duration, usages []certificatesv1.KeyUsage, privateKey interface{}) (reqName string, reqUID types.UID, err error) {
```
make this a pointer so it's clear to callers it is optional

liggitt commented on staging/src/k8s.io/client-go/util/certificate/csr/csr.go:
```diff
@@ -307,6 +310,8 @@ func ensureCompatible(new, orig *certificatesv1.CertificateSigningRequest, priva
 	if len(new.Spec.SignerName) > 0 && len(orig.Spec.SignerName) > 0 && new.Spec.SignerName != orig.Spec.SignerName {
 		return fmt.Errorf("csr signerNames differ: new %q, orig: %q", new.Spec.SignerName, orig.Spec.SignerName)
 	}
+	// TODO do we need a check for durationHint here?
```
we definitely have to tolerate the field not being present to avoid breaking against old servers. given the signer doesn't have to honor the hint, I would tend not to include it in our "should I start a new CSR" logic

kikisdeliveryservice commented:
For ref on this change: https://github.com/kubernetes/enhancements/pull/2759#issuecomment-848156775

There is no active kep for this and we are also past enhancements freeze so an exception will be required if this is to be in 1.22.

kikisdeliveryservice commented:
This has a pending exception request . Removing from milestone until exception is reviewed and KEP is finished.

/hold

enj committed:
Generated

Signed-off-by: Monis Khan <mok@vmware.com>

kikisdeliveryservice commented:
Removing hold since the [KEP has merged](https://github.com/kubernetes/enhancements/pull/2788) and was reviewed by enhancements team

/hold cancel

enj commented:
/milestone v1.22

enj committed:
csr: add expirationSeconds field to control cert lifetime

This change updates the CSR API to add a new, optional field called
expirationSeconds.  This field is a request to the signer for the
maximum duration the client wishes the cert to have.  The signer is
free to ignore this request based on its own internal policy.  The
signers built-in to KCM will honor this field if it is not set to a
value greater than --cluster-signing-duration.  The minimum allowed
value for this field is 600 seconds (ten minutes).

This change will help enforce safer durations for certificates in
the Kube ecosystem and will help related projects such as
cert-manager with their migration to the Kube CSR API.

Future enhancements may update the Kubelet to take advantage of this
field when it is configured in a way that can tolerate shorter
certificate lifespans with regular rotation.

Signed-off-by: Monis Khan <mok@vmware.com>

enj committed:
csr: update e2e conformance test with expirationSeconds usage

Signed-off-by: Monis Khan <mok@vmware.com>

enj commented:
@liggitt this PR is up to date with the KEP now.  The last commit (099123244e8f450f4d3be1b9c258c9457c1adfdf) updates the e2e test, which I think is wrong for this functionality while it is in beta.  I can drop that commit and add it once this is GA if you think that is more correct.

enj commented:
@RainbowMango @logicalhan do y'all have any advice on how to collect metrics duration an integration test?  Gathering from the global metrics registry does not work even if all metrics are unhidden.  `/metrics` does not include them either.  They are available when running a full cluster so I think I got the wiring correct.

liggitt commented on pkg/printers/internalversion/printers.go:
```diff
@@ -412,6 +413,7 @@ func AddHandlers(h printers.PrintHandler) {
 		{Name: "Name", Type: "string", Format: "name", Description: metav1.ObjectMeta{}.SwaggerDoc()["name"]},
 		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
 		{Name: "SignerName", Type: "string", Description: certificatesv1beta1.CertificateSigningRequestSpec{}.SwaggerDoc()["signerName"]},
+		{Name: "RequestedDuration", Type: "string", Description: certificatesv1beta1.CertificateSigningRequestSpec{}.SwaggerDoc()["expirationSeconds"]},
```
nit: move this after the username (who requested it is more important than how long it was requested for)

liggitt commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -344,7 +344,8 @@ func requestNodeCertificate(ctx context.Context, client clientset.Interface, pri
 	}
 
-	reqName, reqUID, err := csr.RequestCertificate(client, csrData, name, certificatesv1.KubeAPIServerClientKubeletSignerName, usages, privateKey)
+	var duration *time.Duration // TODO: pick a good value for the CSR expirationSeconds field and/or allow it to be configured
```
drop the todo, I'd expect a cluster-level kubelet signer duration rather than per-request config like this

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+	namespace = "apiserver"
+	subsystem = "certificates_registry"
+)
+
```
will defer to sig-instrumentation reviewer for this file

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+				return
+			}
+
+			oldCSR := old.(*certificates.CertificateSigningRequest)
```
```suggestion
			oldCSR, ok := old.(*certificates.CertificateSigningRequest)
			if !ok {
				return
			}
```

enj replied:
By the time this function runs, the all of the strategy funcs have already run and they do the exact same casts so not sure why I would not do them here?

liggitt replied:
I'm just not as familiar with all the edge cases where this function is called and I didn't want a panic here

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+				return // ignore CSRs that are not using the CSR duration feature
+			}
+
+			issuedCert := obj.(*certificates.CertificateSigningRequest).Status.Certificate
```
```suggestion
			csr, ok := obj.(*certificates.CertificateSigningRequest)
			if !ok {
				return
			}
			issuedCert := csr.Status.Certificate
```

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "csr_duration_requested",
+			Help:           "Total number of issued CSRs with a requested duration, sliced by signer",
```
clarify that only kubernetes.io signer names are specifically identified?

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "csr_duration_honored",
+			Help:           "Total number of issued CSRs with a requested duration that was honored, sliced by signer",
```
clarify that only kubernetes.io signer names are specifically identified?

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+			}
+
+			// now we check to see if the signer honored the requested duration
+			certificate := certs[0]
```
does ParseCertsPEM always return a non-zero-length list?

enj replied:
Yes

https://github.com/kubernetes/kubernetes/blob/657d93c4cc5201a9fc8a5ba1258d05d81ef39cbd/staging/src/k8s.io/client-go/util/cert/pem.go#L58-L60

I can add a guard statement if you want?

liggitt replied:
no, it's ok

liggitt commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+
+	// short-lived cert backdating + 5% of want
+	// TODO should we have an upper limit on the 5%?
+	maxDelta := 5*time.Minute + (want / 20)
```
I see the 5 minutes from our hard-coded backdate... where does the 5% come from?

enj replied:
I was trying to come up with some fudge factor to allow for external signers that may do like a 10 minute backdate or something.  I can drop it all together if you want...

liggitt commented on pkg/controller/certificates/authority/policies.go:
```diff
@@ -44,14 +47,18 @@ type SigningPolicy interface {
 	TTL time.Duration
 
+	// ExpirationSeconds is used in certificate NotAfter calculation as described above.
+	ExpirationSeconds *int32
```
I'd suggest leaving PermissiveSigningPolicy alone and calculating the TTL of the cert before constructing PermissiveSigningPolicy. That saves you plumbing and teaching more layers about feature enablement

liggitt commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -185,17 +185,18 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
-		Backdate: 5 * time.Minute, // this must always be less than the minimum TTL requested by a user
-		Short:    8 * time.Hour,   // 5 minutes of backdating is roughly 1% of 8 hours
-		Now:      now,
+		TTL:               s.certTTL,
```
update docs on certTTL to indicate this is the max TTL and that individual requests might request shorter certs

liggitt commented on pkg/registry/certificates/certificates/storage/storage.go:
```diff
@@ -62,6 +62,7 @@ func NewREST(optsGetter generic.RESTOptionsGetter) (*REST, *StatusREST, *Approva
 	statusStore := *store
 	statusStore.UpdateStrategy = csrregistry.StatusStrategy
 	statusStore.ResetFieldsStrategy = csrregistry.StatusStrategy
+	statusStore.BeginUpdate = countCSRDurationMetric(csrDurationRequested, csrDurationHonored)
```
hurggggggg

enj replied:
🤣 open to alternatives

liggitt replied:
the characteristics you wanted were "access to old and new object", and "guaranteed to run at most once per persisted status update that issues the cert?"

I don't see another place for both of those :-/

logicalhan commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+			Help:           "Total number of issued CSRs with a requested duration that was honored, sliced by signer",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"signerName"},
```
what's the cardinality on this?

enj replied:
It gets [compressed](https://github.com/kubernetes/kubernetes/pull/99494/files#diff-64f68be0f72515206a7189e4688c140293bb7f36a18a55d9edc001833c42a408R164-R170) to about `O(5)`.

enj replied:
These are the Kube signers:

https://github.com/kubernetes/kubernetes/blob/657d93c4cc5201a9fc8a5ba1258d05d81ef39cbd/staging/src/k8s.io/api/certificates/v1beta1/types.go#L119-L139

logicalhan replied:
that's fine, I was concerned it was unbounded.

logicalhan commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,170 @@
+
+var registerMetricsOnce sync.Once
+
+func TestCollectMetrics(ch chan<- prometheus.Metric) {
```
I'd actually just make the metrics public rather than do this..

enj replied:
The `prometheus` import is forbidden via CI and while I could add an exemption, I would prefer to find a better way to get the metrics data.

logicalhan reviewed (commented):
> @RainbowMango @logicalhan do y'all have any advice on how to collect metrics duration an integration test? Gathering from the global metrics registry does not work even if all metrics are unhidden. /metrics does not include them either. They are available when running a full cluster so I think I got the wiring correct.

Which test are you referring to?

enj commented:
> > @RainbowMango @logicalhan do y'all have any advice on how to collect metrics duration an integration test? Gathering from the global metrics registry does not work even if all metrics are unhidden. /metrics does not include them either. They are available when running a full cluster so I think I got the wiring correct.
> 
> Which test are you referring to?

@logicalhan `TestCSRDuration` in `test/integration/certificates/duration_test.go`.  See the logic near the beginning that tries to assert metrics generated by the test data.

enj commented:
> > > @RainbowMango @logicalhan do y'all have any advice on how to collect metrics duration an integration test? Gathering from the global metrics registry does not work even if all metrics are unhidden. /metrics does not include them either. They are available when running a full cluster so I think I got the wiring correct.
> > 
> > 
> > Which test are you referring to?
> 
> @logicalhan `TestCSRDuration` in `test/integration/certificates/duration_test.go`. See the logic near the beginning that tries to assert metrics generated by the test data.

@logicalhan actually never mind, I figured out what I was doing wrong (the tests were running in parallel and I was not waiting for them correctly).

enj commented:
Most comments [addressed](https://github.com/kubernetes/kubernetes/compare/319d942d2f468f44190fbf8eb4e6117a2a7cdf25..bbf2096864b87e69b268db2af762ff254dcc4a55).

enj commented:
/test all

liggitt commented on test/e2e/auth/certificates.go:
```diff
@@ -205,9 +214,10 @@ var _ = SIGDescribe("Certificates API [Privileged:ClusterAdmin]", func() {
-				Usages:     []certificatesv1.KeyUsage{certificatesv1.UsageDigitalSignature, certificatesv1.UsageKeyEncipherment, certificatesv1.UsageServerAuth},
+				Request:           csrData,
+				SignerName:        signerName,
+				ExpirationSeconds: csr.DurationToExpirationSeconds(time.Hour),
```
I think this should also be a TODO... no sense setting beta data we're not verifying in a conformance test

liggitt commented:
API change lgtm, one comment on the conformance TODO

I think the desired TTL for the certificate should be computed in the sign() method, rather than plumbing down one more layer. The signer having a field for the maxTTL and a sign() parameter for the requested instance ttl makes sense... it should manage joining those and propagate the result into the policy TTL field

liggitt commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -203,6 +206,24 @@ func (s *signer) sign(x509cr *x509.CertificateRequest, usages []capi.KeyUsage, n
+
+	// honor requested duration is if it is less than the default TTL
+	// use 10 min (2x hard coded backdate above) as a sanity check lower bound
+	if requestedDuration := csr.ExpirationSecondsToDuration(*expirationSeconds); requestedDuration < s.certTTL && requestedDuration >= 10*time.Minute {
```
if this is less than our lower bound we ignore it, rather than clip to our lower bound?

enj replied:
Let me split the two cases.  It does not matter today but it could in the future if we changed API validation.

enj commented:
@logicalhan @RainbowMango could you please review the metrics?

https://github.com/kubernetes/kubernetes/pull/99494#discussion_r659228781

logicalhan commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,163 @@
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "csr_duration_requested",
```
`csr_requested_duration_total` would be more conventional
/cc @brancz to verify

enj replied:
Fixed.

logicalhan commented on pkg/registry/certificates/certificates/storage/metrics.go:
```diff
@@ -0,0 +1,163 @@
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "csr_duration_honored",
```
same here `csr_honored_duration_total`

enj replied:
Fixed.

liggitt commented:
/retest

liggitt commented:
/lgtm
/approve

enj commented:
/retest

enj commented:
/retest
